pkg/resource/autonaming/strategy.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autonaming
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/urn"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+)
+
+// autonamingStrategy is a strategy for autonaming a resource.
+type autonamingStrategy interface {
+	GetOptions(urn urn.URN, randomSeed []byte) (opts *plugin.AutonamingOptions, deleteBeforeCreate bool)
+}
```

Comment:
Do any of the values in `AutonamingOptions` require that `deleteBeforeCreate` is true or false to work correctly? I'm wondering if `deleteBeforeCreate` could be inferred from the options, or if it should be part of the options struct?

Reply:
It's mostly impossible to tell if `deleteBeforeCreate` is required based on other options (proposed name and mode, except for disabled).

Initially I had `deleteBeforeCreate` as an extra option, but it's a flag for the step generator, not for a provider plugin, so it felt wrong to include it there?

`WarnIfNoSupport` is another odd one because it doesn't get to the provider's gRPC but is needed for provider plugin because that's the point where we know the ConfigureResult/HandshakeResult/getPluginConfig result.

Any ideas how to make it cleaner?

Reply:
As long as you've kept a "random" portion then deleteBeforeCreate isn't needed. Plus there is the odd resource that will allow a duplicate name, so I'm not sure we want to try and limit that in the engine?

Reply:
@Frassle A lot of names (verbatim, disabled, pattern without a random expression) have no randomness. The provider has no way of figuring out if a resource name is random or not either. I think ~all resources that we care if they are delete-before-replace or create-before-replace won't allow duplicate names.